handlers/idempotency.go

```diff
@@ -0,0 +1,207 @@
+	ExpiryDate time.Time `gorm:"column:expiry_date"`
+}
+
+func (IdempotencyStoreGormItem) TableName() string {
+	return "idempotency_keys"
+}
+
+func NewIdempotencyStoreGorm(db *gorm.DB) *IdempotencyStoreGorm {
+	return &IdempotencyStoreGorm{db: db}
+}
+
+func (g *IdempotencyStoreGorm) Get(key string) (bool, error) {
+	item := IdempotencyStoreGormItem{}
+	err := g.db.First(&item, "key = ? and expiry_date > ?", key, time.Now()).Error
+	if errors.Is(err, gorm.ErrRecordNotFound) {
+		// key doesn't exist
+		return false, nil
+	} else if err != nil {
+		// some other error
+		return false, err
+	}
+
+	// key exists
+	return true, nil
+}
+
+func (g *IdempotencyStoreGorm) Set(key string, expiry time.Duration) error {
+	// update expiry date if exists or create a new item
+	err := g.db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "key"}},
+		DoUpdates: clause.AssignmentColumns([]string{"expiry_date"}),
+	}).Create(&IdempotencyStoreGormItem{Key: key, ExpiryDate: time.Now().Add(expiry)}).Error
+
+	if err != nil {
+		fmt.Println("error when creating IdempotecyStoreGormItem:", err)
+		return err
+	}
+
+	return nil
+}
+
+// Prune deletes all expired IdempotencyStoreGormItems from the database
+func (g *IdempotencyStoreGorm) Prune() error {
+	err := g.db.Delete(IdempotencyStoreGormItem{}, "expiry_date < ?", time.Now()).Error
+	return err
+}
+
+// Local / in-memory store for idempotency keys, mainly for testing purposes
+type IdempotencyStoreLocal struct {
+	keys map[string]time.Time // key: expiry
+}
+
+func NewIdempotencyStoreLocal() *IdempotencyStoreLocal {
+	return &IdempotencyStoreLocal{make(map[string]time.Time)}
+}
+
+func (m *IdempotencyStoreLocal) Get(key string) (bool, error) {
+	v, ok := m.keys[key]
+	if !ok {
+		return false, nil
+	}
+
+	// Still valid
+	if v.After(time.Now()) {
+		return true, nil
+	}
+
+	// Expired
+	// NOTE: item is removed as a side effect
+	if v.Before(time.Now()) {
+		delete(m.keys, key)
+		return false, nil
+	}
+
+	return false, nil
+}
+
+func (m *IdempotencyStoreLocal) Set(key string, expiry time.Duration) error {
+	dl := time.Now().Add(expiry)
+	m.keys[key] = dl
+
+	return nil
+}
+
+// IdempotencyHandler returns a http.HandlerFunc that checks
+// for request idempotency when applicable
+func IdempotencyHandler(h http.Handler, opts IdempotencyHandlerOptions, store IdempotencyStore) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		// Check for ignored paths
+		for _, path := range opts.IgnorePaths {
+			if strings.HasPrefix(r.URL.Path, path) {
+				h.ServeHTTP(rw, r)
+				return
+			}
+		}
+
+		// Only POST requests are checked
+		if r.Method != http.MethodPost {
+			h.ServeHTTP(rw, r)
+			return
+		}
+
+		key := r.Header.Get("Idempotency-Key")
+		if len(key) == 0 && r.Method == http.MethodPost {
+			http.Error(rw, "Idempotency-Key header not found", http.StatusBadRequest)
+			return
+		}
+
+		exists, err := store.Get(key)
+		if err != nil {
+			panic(err)
```

Comment:
Are we sure we want to panic in a http handler?

Reply:
good catch :D no we most certainly don't!